other/no-loss-lottery/staking/src/lib.rs

```diff
@@ -0,0 +1,185 @@
+        }
+
+        // Calculates reward per token in supply
+        fn get_reward_value(&self) -> Decimal {
+
+            if self.staking_pool.is_empty() {
+                debug!("Pool is empty");
+                return Decimal::zero();
+            }
+
+            // calculate per token reward depending on the time passed since last epoch
+            let time = Context::current_epoch()-self.last_update;
+            let exp = 1e18 as i64;
+            let exp_rate = Decimal::from(exp) / self.staking_pool.amount();
+            let epoch_rate = self.rate / 10000;
+            let calculation = Decimal::from(time) * epoch_rate * exp_rate;
+            let result = self.reward_value + calculation;
+            result
+        }
+
+        // calculate earning 
+        fn earned(&self, account: Address) -> Decimal {
+
+            let exp = 1e18 as i64;
+            let reward_value = self.get_reward_value();
+
+            // get account data
+            let balance = match self.balances.get(&account) {
+                Some(value) => *value,
+                None => Decimal::zero()
+            };
+
+            debug!("Account balance: {}", balance);
+
+            let reward = match self.rewards.get(&account){
+                Some(value) => *value,
+                None => Decimal::zero()
+            };
+
+            debug!("Account reward: {}", reward);
+
+            let paid = match self.paid.get(&account){
+                Some(value) => *value,
+                None => Decimal::zero()
+            };
+
+            // calculate reward per account by balance current reward and collected paid reward
+            let calculation = (balance * (reward_value - paid)) / Decimal::from(exp);
+
+            debug!("Earned calculation: {}", calculation);
+
+            let result = reward + calculation;
+
+            debug!("Earned result: {}", result);
+
+            result 
+        }
+     
+        // private method that is used to update rewards during the epoch
+        // will be called on each public request
+        fn update_reward(&mut self, account: Address) {
+            // update per token reward
+            self.reward_value = self.get_reward_value();
+            // update epoch
+            self.last_update = Context::current_epoch();
+            // calculate earnings
+            let earned = self.earned(account);
+            // store earnings for the account
+            self.rewards.insert(account, earned);
+            // set updated per token reward value to the account
+            self.paid.insert(account, self.reward_value);
+        }
+
+        /// Registers a new user
+        pub fn new_user(&self) -> Bucket {
+            ResourceBuilder::new_fungible(DIVISIBILITY_NONE)
+                .metadata("name", "Staking User Badge")
+                .initial_supply_fungible(1)
+        }
+
+        // staking for specific account
+        pub fn stake(&mut self, user_auth: BucketRef, bucket: Bucket){
```

Comment:
Here, instead of having the user_auth in the params, you could specify it with `#[auth(user_nft_resource_def)]`. Assuming you did the changes to use NFTs instead.

Reply:
Hey @cbisaillon. I'm not sure I understand the solution. Let's say I made this for registering a user:

```
pub fn new_user(&self) -> Bucket {
            // mint course NFT
            let token_data = StakingData {
                reward: Decimal::zero(),
                balance: Decimal::zero(),
                paid: Decimal::zero()
            };

            let bucket = self.token_minter.authorize(|auth| {
                self.token_resource_def
                    .mint_nft(self.token_id_counter, token_data, auth)
            });
            self.token_id_counter += 1;
            
            bucket
        }
``` 

So now the user has an NFT token, then how's the #auth will work if I need to update the user's NFT metadata?
I thought that I should put this NFT in addition to the staking amount so its metadata will be updated and the NFT will be returned back to the user after the update. Is the following code a valid pattern?

 ```
pub fn stake(&mut self, token: Bucket, staking: Bucket) -> Bucket {

             // get nft data from token
            let id = token.get_nft_id();
            let mut data: StakingData = token.get_nft_data(id);

            let reward = self.update_reward(data);
            let amount = staking.amount();

            data.reward = reward;
            data.paid = self.reward_value;
            data.balance += amount;

            self.token_minter
            .authorize(|auth| token.update_nft_data(id, data, auth));

            debug!("Account staking balance: {}", data.balance);

            self.staking_pool.put(staking);

            token
        }
```

Reply:
Then, instead of doing `token.update_nft_data`, you would use the resource_def: `self.nft_resource_def.update_nft_data()`

Reply:
Wow, this is cool. Working on the update, thank you!